x-pack/legacy/plugins/apm/server/lib/apm_telemetry/index.ts

```diff
@@ -4,66 +4,109 @@
+import { CoreSetup, Logger } from 'src/core/server';
+import { Observable } from 'rxjs';
+import { take } from 'rxjs/operators';
+import { PluginSetupContract as TaskManager } from '../../../../task_manager/plugin';
+import { APMConfig } from '../../../../../../plugins/apm/server';
 import { getInternalSavedObjectsClient } from '../helpers/saved_objects_client';
-import {
-  APM_SERVICES_TELEMETRY_SAVED_OBJECT_TYPE,
-  APM_SERVICES_TELEMETRY_SAVED_OBJECT_ID
-} from '../../../common/apm_saved_object_constants';
 import { APMLegacyServer } from '../../routes/typings';
 import { UsageCollectionSetup } from '../../../../../../../src/plugins/usage_collection/server';
+import { getApmIndices } from '../settings/apm_indices/get_apm_indices';
+import {
+  APM_TELEMETRY_SAVED_OBJECT_ID,
+  APM_TELEMETRY_SAVED_OBJECT_TYPE
+} from '../../../common/apm_saved_object_constants';
+import {
+  collectDataTelemetry,
+  CollectTelemetryParams
+} from './collect_data_telemetry';
 
-export function createApmTelementry(
-  agentNames: string[] = []
-): SavedObjectAttributes {
-  const validAgentNames = agentNames.filter(isAgentName);
-  return {
-    has_any_services: validAgentNames.length > 0,
-    services_per_agent: countBy(validAgentNames)
-  };
-}
+const APM_TELEMETRY_TASK_NAME = 'apm-telemetry-task';
+
+export function createApmTelemetry({
+  core,
+  config$,
+  server,
+  usageCollector,
+  taskManager,
+  logger
+}: {
+  core: CoreSetup;
+  config$: Observable<APMConfig>;
+  server: APMLegacyServer;
+  usageCollector: UsageCollectionSetup;
+  taskManager: TaskManager;
+  logger: Logger;
+}) {
+  const savedObjectsClient = getInternalSavedObjectsClient(server);
+
+  const collectAndStore = async () => {
+    const [esClient, config] = await Promise.all([
+      core.elasticsearch.dataClient$.pipe(take(1)).toPromise(),
+      config$.pipe(take(1)).toPromise()
+    ]);
+
+    const indices = await getApmIndices({
+      config,
+      savedObjectsClient: getInternalSavedObjectsClient(server)
+    });
+
+    const search = esClient.callAsInternalUser.bind(
+      null,
+      'search'
+    ) as CollectTelemetryParams['search'];
+
+    const indicesStats = esClient.callAsInternalUser.bind(
+      null,
+      'indices.stats'
+    ) as CollectTelemetryParams['indicesStats'];
+
+    const dataTelemetry = await collectDataTelemetry({
+      search,
+      indices,
+      logger,
+      indicesStats
+    });
 
-export async function storeApmServicesTelemetry(
-  server: APMLegacyServer,
-  apmTelemetry: SavedObjectAttributes
-) {
-  try {
-    const savedObjectsClient = getInternalSavedObjectsClient(server);
     await savedObjectsClient.create(
-      APM_SERVICES_TELEMETRY_SAVED_OBJECT_TYPE,
-      apmTelemetry,
-      {
-        id: APM_SERVICES_TELEMETRY_SAVED_OBJECT_ID,
-        overwrite: true
-      }
+      APM_TELEMETRY_SAVED_OBJECT_TYPE,
+      dataTelemetry,
+      { id: APM_TELEMETRY_SAVED_OBJECT_TYPE, overwrite: true }
```

Comment:
There is a risk we might delete existing telemetry because of 
https://github.com/elastic/kibana/blob/caee49b307bf9a6c8c1c146b1a7c35fd044fd894/x-pack/legacy/plugins/apm/server/lib/apm_telemetry/collect_data_telemetry/index.ts#L64-L68

But the chances should be low so I wouldn't worry too much about it (just a heads up) ;)